File: example/example-laz-writer.cpp
#include <cassert>
#include <random>

#include <copc-lib/geometry/vector3.hpp>
#include <copc-lib/io/laz_writer.hpp>

using namespace copc;
using namespace std;

// constants
const Vector3 MIN_BOUNDS = {-2000, -5000, 20}; // Scaled coordinates
const Vector3 MAX_BOUNDS = {5000, 1034, 125};  // Scaled coordinates
const int NUM_POINTS = 3000;

// random num devices
std::random_device rd;  // obtain a random number from hardware
std::mt19937 gen(rd()); // seed the generator

// This function will generate `NUM_POINTS` random points within the LAS bounds
las::Points RandomPoints(const las::LasHeader &header, int number_points)
{

    // Random num generators between the min and max spatial bounds of the las header
    std::uniform_int_distribution<> rand_x(header.RemoveScaleX(header.min.x), header.RemoveScaleX(header.max.x));
    std::uniform_int_distribution<> rand_y(header.RemoveScaleY(header.min.y), header.RemoveScaleY(header.max.y));
    std::uniform_int_distribution<> rand_z(header.RemoveScaleZ(header.min.z), header.RemoveScaleZ(header.max.z));

    // Create a Points object based on the LAS header
    las::Points points(header);

    // Populate the points
    for (int i = 0; i < number_points; i++)
    {
        // Create a point with a given point format
        // The use of las::Point constructor is strongly discouraged, instead use las::Points::CreatePoint
        auto point = points.CreatePoint();
        // point has getters/setters for all attributes
        point->X(header.ApplyScaleX(rand_x(gen)));
        point->Y(header.ApplyScaleY(rand_y(gen)));
        point->Z(header.ApplyScaleZ(rand_z(gen)));

        points.AddPoint(point);
    }
    return points;
}

// In this example, we'll create our own LAZ file from scratch
void NewFileExample()
{
    // Create our new LAZ file with the specified format, scale, offset, and wkt
    las::LazConfigWriter cfg(8, {0.1, 0.1, 0.1}, {50, 50, 50}, "TEST_WKT");
    // copc-lib will not automatically compute the min/max of added points
    // so we will have to calculate it ourselves
    cfg.LasHeader()->min = MIN_BOUNDS;
    cfg.LasHeader()->max = MAX_BOUNDS;

    // Now, we can create a LAZ writer
    laz::LazFileWriter writer("new-laz.laz", cfg);
    auto header = *writer.LazConfig()->LasHeader();

    // Let's generate some points
    auto points = RandomPoints(header, NUM_POINTS);

    // The points will be written to the file when we call WritePoints
    writer.WritePoints(points);

    // We can write a second batch of points in a different chunk
    points = RandomPoints(header, NUM_POINTS);
    writer.WritePoints(points);

    // Make sure we call close to finish writing the file!
    writer.Close();
}

int main() { NewFileExample(); }

File: example/example_laz_writer.py
import random
import copclib as copc
import os

random.seed(0)
DATADIRECTORY = os.path.join(os.path.dirname(__file__), "..", "test", "data")
if not os.path.exists(os.path.join(os.path.join(DATADIRECTORY, "out"))):
    os.makedirs(os.path.join(DATADIRECTORY, "out"))

MIN_BOUNDS = copc.Vector3(-2000, -5000, 20)  # Argument Constructor
MAX_BOUNDS = copc.Vector3([5000, 1034, 125])  # List Constructor
NUM_POINTS = 3000


# This function will generate `NUM_POINTS` random points within the voxel bounds
def RandomPoints(las_header, number_points):

    # Create a Points object based on the LAS header
    points = copc.Points(las_header)

    # Populate the points
    for i in range(number_points):
        # Create a point with a given point format
        point = points.CreatePoint()
        # point has getters/setters for all attributes
        point.x = random.uniform(
            las_header.min.x,
            las_header.max.x,
        )
        point.y = random.uniform(
            las_header.min.y,
            las_header.max.y,
        )
        point.z = random.uniform(
            las_header.min.z,
            las_header.max.z,
        )

        points.AddPoint(point)
    return points


def writer_example():
    # In this example, we'll create our own file from scratch

    # Create our new file with the specified format, scale, and offset
    cfg = copc.LazConfigWriter(
        point_format_id=8,
        scale=(0.1, 0.1, 0.1),
        offset=(50, 50, 50),
        wkt="TEST_WKT",
    )
    # As of now, the library will not automatically compute the min/max of added points
    # so we will have to calculate it ourselves
    cfg.las_header.min = MIN_BOUNDS
    cfg.las_header.max = MAX_BOUNDS

    # Now, we can create our COPC writer:
    writer = copc.LazWriter(os.path.join(DATADIRECTORY, "out", "test-laz.laz"), cfg)

    header = writer.laz_config.las_header

    points = RandomPoints(header, NUM_POINTS)
    writer.WritePoints(points)

    points = RandomPoints(header, NUM_POINTS)
    writer.WritePoints(points)

    writer.Close()

    # validate using the reader
    reader = copc.LazReader(os.path.join(DATADIRECTORY, "out", "test-laz.laz"))
    assert reader.laz_config.las_header.point_count == NUM_POINTS * 2
    assert len(reader.GetPoints()) == NUM_POINTS * 2
    print("DONE")


if __name__ == "__main__":
    writer_example()
